File: Sources/redline/Alerts.swift
// Delivery for the events Alerting decides on. The decision is in the core and tested; this
// is the part that talks to macOS.
//
// Nothing is posted until the user turns alerts on, and the authorization prompt is raised at
// that moment rather than on first launch: an app that asks to notify before it has anything
// to say is an app people say no to.
import Foundation
import RedlineCore
import UserNotifications

final class AlertCenter {
    private var state = AlertStore.load()
    private var cadenceState = CadenceStore.load()
    /// Notifications are unavailable outside an app bundle, which is how the binary runs
    /// from a plain `swift build`. Checked once rather than guarded at every call site.
    private let available = Bundle.main.bundleIdentifier != nil
    private var authorized = false
    private var askedThisLaunch = false

    /// Posts whatever this poll's readings have newly earned. Returns the events fired, so
    /// the caller can log or test without reaching into the notification centre.
    @discardableResult
    func evaluate(windows: [LimitWindow], paces: [Pace], config: Config,
                  isStale: @escaping (LimitWindow) -> Bool = { _ in false },
                  now: Date = Date()) -> [AlertEvent] {
        let events = Alerting.evaluate(windows: windows, paces: paces, config: config,
                                       now: now, isStale: isStale, state: &state)
        AlertStore.save(state)
        guard config.alerts, !events.isEmpty else { return events }
        deliver(events)
        return events
    }

    /// Posts whatever the shape of the day has newly earned. Same delivery path and same
    /// permission as the limit alerts; a cue never makes a sound, because none of them is
    /// the kind of thing that should pull someone out of what they are doing.
    @discardableResult
    func evaluateCadence(entries: [Entry], config: Config,
                         now: Date = Date()) -> [CadenceCue] {
        let cues = CadenceRules.evaluate(entries: entries, config: config, now: now,
                                         state: &cadenceState)
        CadenceStore.save(cadenceState)
        guard config.mindfulCues, !cues.isEmpty, available else { return cues }
        let center = UNUserNotificationCenter.current()
        if !authorized, !askedThisLaunch {
            requestAuthorization { [weak self] granted in
                guard granted else { return }
                self?.post(cues, to: center)
            }
            return cues
        }
        post(cues, to: center)
        return cues
    }

    private func post(_ cues: [CadenceCue], to center: UNUserNotificationCenter) {
        for cue in cues {
            let content = UNMutableNotificationContent()
            content.title = cue.title
            content.body = cue.body
            center.add(UNNotificationRequest(identifier: cue.id, content: content,
                                             trigger: nil))
        }
    }

    /// Called when the setting is switched on. macOS only shows the prompt once per app, so
    /// a refusal is remembered by the system and this simply stops delivering.
    func requestAuthorization(completion: ((Bool) -> Void)? = nil) {
        guard available else {
            completion?(false)
            return
        }
        askedThisLaunch = true
        UNUserNotificationCenter.current()
            .requestAuthorization(options: [.alert, .sound]) { [weak self] granted, _ in
                DispatchQueue.main.async {
                    self?.authorized = granted
                    completion?(granted)
                }
            }
    }

    private func deliver(_ events: [AlertEvent]) {
        guard available else { return }
        let center = UNUserNotificationCenter.current()
        // Authorization is asked for on the first delivery too, in case the setting was
        // turned on by hand in config.json rather than from the menu.
        if !authorized, !askedThisLaunch {
            requestAuthorization { [weak self] granted in
                guard granted else { return }
                self?.post(events, to: center)
            }
            return
        }
        post(events, to: center)
    }

    private func post(_ events: [AlertEvent], to center: UNUserNotificationCenter) {
        for event in events {
            let content = UNMutableNotificationContent()
            content.title = event.title
            content.body = event.body
            // A limit actually reached is the only one that gets a sound; the rest are
            // information arriving while you work.
            if case .limitReached = event.kind { content.sound = .default }
            center.add(UNNotificationRequest(identifier: event.id, content: content,
                                             trigger: nil))
        }
    }
}

File: windows/src/Redline.App/Services/Alerts.cs
// Delivery for the events Alerting decides on. The decision is in the core and tested; this
// is the part that hands them to Windows, through whatever INotifier the tray supplies.
using Redline.Core;

namespace Redline.App.Services;

/// One notification. Sound is reserved for a limit actually reached.
public interface INotifier
{
    void Show(string title, string body, bool sound);
}

/// The tray's NotifyIcon as the notifier. Windows 10 and later turn balloons into toasts.
public sealed class NotifyIconNotifier : INotifier
{
    private readonly System.Windows.Forms.NotifyIcon _icon;
    public NotifyIconNotifier(System.Windows.Forms.NotifyIcon icon) => _icon = icon;

    // The shell truncates past these lengths anyway; cutting here keeps the end readable
    public const int MaxTitle = 63;
    public const int MaxBody = 255;

    public void Show(string title, string body, bool sound)
    {
        var t = title.Length > MaxTitle ? title[..(MaxTitle - 1)] + "…" : title;
        var b = body.Length > MaxBody ? body[..(MaxBody - 1)] + "…" : body;
        // Only a reached limit may make noise; the rest are information arriving while you work
        var icon = sound ? System.Windows.Forms.ToolTipIcon.Warning : System.Windows.Forms.ToolTipIcon.Info;
        _icon.ShowBalloonTip(10_000, t, b, icon);
        if (sound) System.Media.SystemSounds.Exclamation.Play();
    }
}

public sealed class AlertCenter
{
    private readonly object _lock = new();
    private readonly AlertState _state;
    private readonly CadenceState _cadenceState;
    private readonly string? _alertPath;
    private readonly string? _cadencePath;

    /// Null until the tray exists; events are still recorded, so nothing fires twice later.
    public INotifier? Notifier { get; set; }

    public AlertCenter(INotifier? notifier = null, string? alertStorePath = null, string? cadenceStorePath = null)
    {
        Notifier = notifier;
        _alertPath = alertStorePath;
        _cadencePath = cadenceStorePath;
        _state = AlertStore.Load(alertStorePath);
        _cadenceState = CadenceStore.Load(cadenceStorePath);
    }

    /// Posts whatever this poll's readings have newly earned; a stale reading updates the
    /// record but never fires. Returns the events, so the caller can log or test.
    public List<AlertEvent> Evaluate(IEnumerable<LimitWindow> windows, IEnumerable<Pace> paces, Config config,
                                     Func<LimitWindow, bool>? isStale = null, DateTimeOffset? now = null)
    {
        List<AlertEvent> events;
        lock (_lock)
        {
            events = Alerting.Evaluate(windows, config, _state, paces, now, isStale ?? (_ => false));
            AlertStore.Save(_state, _alertPath);
        }
        if (!config.Alerts || events.Count == 0) return events;
        foreach (var e in events) Post(e.Title, e.Body, e.Kind is AlertKind.LimitReached);
        return events;
    }

    /// The shape of the day. Same delivery path; a cue never makes a sound, because none of
    /// them should pull someone out of what they are doing.
    public List<CadenceCue> EvaluateCadence(IReadOnlyList<Entry> entries, Config config, DateTimeOffset? now = null)
    {
        List<CadenceCue> cues;
        lock (_lock)
        {
            cues = CadenceRules.Evaluate(entries, config, _cadenceState, now);
            CadenceStore.Save(_cadenceState, _cadencePath);
        }
        if (!config.MindfulCues || cues.Count == 0) return cues;
        foreach (var c in cues) Post(c.Title, c.Body, sound: false);
        return cues;
    }

    private void Post(string title, string body, bool sound)
    {
        if (Notifier is not { } n) return;
        try { n.Show(title, body, sound); }
        catch (Exception e)
        {
            Diag.Log.Error("alerts.deliver_failed", "could not show a notification", new() { ["error"] = e.Message });
        }
    }
}
